src/jit/hwintrinsiclistxarch.h

```diff
@@ -11,58 +11,58 @@
-HARDWARE_INTRINSIC(AVX2_Add,                    "Add",                          AVX2)
+HARDWARE_INTRINSIC(AVX2_IsSupported,                    "get_IsSupported",              AVX2,           {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
+HARDWARE_INTRINSIC(AVX2_Add,                            "Add",                          AVX2,           {INS_invalid, INS_invalid, INS_paddb,   INS_paddw,   INS_paddd,   INS_paddq},       -1)
 
 //  AES Intrinsics 
-HARDWARE_INTRINSIC(AES_IsSupported,             "get_IsSupported",              AES)
+HARDWARE_INTRINSIC(AES_IsSupported,                     "get_IsSupported",              AES,            {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
 
 //  BMI1 Intrinsics 
-HARDWARE_INTRINSIC(BMI1_IsSupported,            "get_IsSupported",              BMI1)
+HARDWARE_INTRINSIC(BMI1_IsSupported,                    "get_IsSupported",              BMI1,           {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
 
 //  BMI2 Intrinsics 
-HARDWARE_INTRINSIC(BMI2_IsSupported,            "get_IsSupported",              BMI2)
+HARDWARE_INTRINSIC(BMI2_IsSupported,                    "get_IsSupported",              BMI2,           {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
 
 //  FMA Intrinsics 
-HARDWARE_INTRINSIC(FMA_IsSupported,             "get_IsSupported",              FMA)
+HARDWARE_INTRINSIC(FMA_IsSupported,                     "get_IsSupported",              FMA,            {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
 
 //  LZCNT Intrinsics 
-HARDWARE_INTRINSIC(LZCNT_IsSupported,           "get_IsSupported",              LZCNT)
-HARDWARE_INTRINSIC(LZCNT_LeadingZeroCount,      "LeadingZeroCount",             LZCNT)
+HARDWARE_INTRINSIC(LZCNT_IsSupported,                   "get_IsSupported",              LZCNT,          {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
+HARDWARE_INTRINSIC(LZCNT_LeadingZeroCount,              "LeadingZeroCount",             LZCNT,          {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_lzcnt,   INS_lzcnt},       -1)
 
 //  PCLMULQDQ Intrinsics 
-HARDWARE_INTRINSIC(PCLMULQDQ_IsSupported,       "get_IsSupported",              PCLMULQDQ)
+HARDWARE_INTRINSIC(PCLMULQDQ_IsSupported,               "get_IsSupported",              PCLMULQDQ,      {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
 
 //  POPCNT Intrinsics 
-HARDWARE_INTRINSIC(POPCNT_IsSupported,          "get_IsSupported",              POPCNT)
-HARDWARE_INTRINSIC(POPCNT_PopCount,              "PopCount",                    POPCNT)
+HARDWARE_INTRINSIC(POPCNT_IsSupported,                  "get_IsSupported",              POPCNT,         {INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid, INS_invalid},     -1)
```

Comment:
Thanks for the work. I have a incomplete implementation in my local repo that I think it would be better to have a “category” field (e.g., simple intrinsic, helper, memory access, etc.).

Reply:
What is the category field useful for? Also, is it in addition to the instruction array?

Reply:
For example, simple intrinsics can be easily table driven, and r64 intrinsics should throw exceptions on x86, and helpers should be manually treated.

Reply:
Okay, so it's moving a bit of the switch statement to also be table driven, so instead of `case add, case sub, etc`, it would be case simple_intrin`?

Reply:
Yes, exactly.

Reply:
Would it end up being better to have these categorized as 'Helper`, `SIMD_R_R`, `SIMD_R_R_R`, `SIMD_R_R_R_I`, etc?

There are some that are simple (such as the compare instructions) but that have a different emit path (R_R_R_I instead of R_R_R).

Each emit path will need to have fairly consistent containment analysis done, so it may be a better "categorization"/split for the jump table.

Reply:
I agree that a category field will be good.

However, after having implemented most of the SSE intrinsics and having started looking at the containment support, I am pretty sure having them categorized by "supported encoding" would be ideal.

So, instructions like "Add, And, AndNot, etc" should be categorized as "R, R/M" instructions. And instructions like `movlhps` should be categorized as "R, R" instructions.

With this type of categorization, we can still cut out most of the switch table by using `switch (category)` rather than `switch (intrinsicID)`. However, we can also have a `genHWIntrinsic<category>` method which centralizes the operand handling for a particular category of instructions. 

For example, `genHWIntrinsicR_RM(instruction ins, GenTree* op1, GenTree* op2)` (used by Add, And, AndNot, etc) will know that op1 cannot be contained, but op2 can optionally be contained and call `emitIns_R_R_R` or `emitIns_R_R_A` as appropriate. Likewise, `genHWIntrinsic_R_R` (used by MoveHighToLow and MoveLowToHigh) will know that neither op can be contained and can skip those checks entirely.

We can also use the category check in lowerxarch to mark nodes as contained or registers as optional.

Thoughts?